server/command.go

```diff
@@ -320,3 +339,125 @@ func (p *Plugin) executeDisconnectBotCommand(args *model.CommandArgs) (*model.Co
+		return nil, nil
+	}
+
+	msUsersMap := make(map[string]msteams.User, len(msChannelMembers))
+	for _, u := range msChannelMembers {
+		msUsersMap[u.Mail] = u
+	}
+
+	return msChannelMembers, msUsersMap
+}
+
+func (p *Plugin) getMattermostUsersAndMap(channelID string) ([]*model.User, map[string]*model.User) {
+	mmChannelMembers, getErr := p.API.GetChannelMembers(channelID, 0, math.MaxInt32)
+	if getErr != nil {
+		p.API.LogError("Unable to get Mattermost users to add in the channel", "error", getErr)
+		return nil, nil
+	}
+
+	var mmUsers []*model.User
+	mmUsersMap := make(map[string]*model.User, len(mmChannelMembers))
+	for _, u := range mmChannelMembers {
+		user, err := p.API.GetUser(u.UserId)
+		if err != nil {
+			p.API.LogError("Unable to get the user", "error", err)
+			continue
+		}
+
+		mmUsers = append(mmUsers, user)
+		mmUsersMap[user.Email] = user
+	}
+
+	return mmUsers, mmUsersMap
+}
+
+func (p *Plugin) addMattermostUser(wg *sync.WaitGroup, teamID, channelID string, msChannelMembers []msteams.User, mmUsersMap map[string]*model.User, batchSize int) {
+	for i := 0; i < (len(msChannelMembers)/batchSize)+1; i++ {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			for j := i * batchSize; j < getMin((i*batchSize)+batchSize, len(msChannelMembers)); j++ {
+				u := msChannelMembers[j]
+				_, ok := mmUsersMap[u.Mail]
+				if !ok {
+					mmUserID, err := p.store.TeamsToMattermostUserID(u.ID)
+					if err != nil {
+						userUUID := uuid.Parse(u.ID)
+						encoding := base32.NewEncoding("ybndrfg8ejkmcpqxot1uwisza345h769").WithPadding(base32.NoPadding)
+						shortUserID := encoding.EncodeToString(userUUID)
+						username := slug.Make(u.DisplayName) + "_" + u.ID
+
+						newUser, appErr := p.API.CreateUser(&model.User{
+							Password:  utils.GenerateRandomPassword(),
+							Email:     u.Mail,
+							RemoteId:  &shortUserID,
+							FirstName: u.DisplayName,
+							Username:  username,
+						})
+
+						if appErr != nil {
+							p.API.LogError("Unable to create new user", "error", appErr)
+							continue
+						}
+
+						setErr := p.store.SetUserInfo(newUser.Id, u.ID, nil)
+						if setErr != nil {
+							p.API.LogError("Unable to store the user", "error", setErr)
+							continue
+						}
+					}
+
+					if _, err := p.API.GetTeamMember(teamID, mmUserID); err != nil {
+						if _, addErr := p.API.CreateTeamMember(teamID, mmUserID); addErr != nil {
+							p.API.LogError("Unable to add user to the team", "error", addErr)
+							continue
+						}
+					}
+
+					if _, err := p.API.AddChannelMember(channelID, mmUserID); err != nil {
+						p.API.LogError("Unable to add user to the channel", "error", err)
+					}
+				}
+			}
+		}(i)
+	}
+}
+
+func (p *Plugin) addMSTeamsUser(wg *sync.WaitGroup, client msteams.Client, teamID, channelID string, mmUsers []*model.User, msUsersMap map[string]msteams.User, batchSize int) {
+	for i := 0; i < (len(mmUsers)/batchSize)+1; i++ {
+		wg.Add(1)
+		go func(i int) {
```

Comment:
Same here, unbounded number of goroutines :(

Reply:
I this case is even worse, we hare hammering the msteams API like crazy. In this case we should do something like batching, at least.